enforcer/src/scheduler/task.c

```diff
@@ -66,20 +107,46 @@ time_t task_execute(task_t *task, db_connection_t *dbconn)
      * really tell the difference between an error and nodata.) Once we
      * fixed our database backend this lock can be removed.
      * */
-    if (!strcmp(task->class, TASK_CLASS_ENFORCER))
-        pthread_mutex_lock(&worklock);
-
+    if(task->lock) {
         pthread_mutex_lock(task->lock);
-            t = task->callback(task->owner, task->context, dbconn);
+            t = task->callback(task->owner, task->userdata, context);
         pthread_mutex_unlock(task->lock);
-
-    if (!strcmp(task->class, TASK_CLASS_ENFORCER))
-        pthread_mutex_unlock(&worklock);
+    } else {
+        t = task->callback(task->owner, task->userdata, context);
+    }
 
     return t;
 }
 
-static int cmp_ttuple(task_t *x, task_t *y)
+void
+task_perform(schedule_type* scheduler, task_type* task, void* context)
```

Comment:
I don't like this design. Only because of this function task.c must be aware of what a scheduler is. A better place would be in the scheduler as shed_execute_and_reschedule_task() or something of your liking.